src/base/memory_monitor_sr.cpp

```diff
@@ -20,18 +20,57 @@
  * memory_monitor_sr.cpp - Implementation of memory monitor module
  */
 
+#include <malloc.h>
+#include <cassert>
 #include <cstring>
 #include <algorithm>
-#include <cassert>
 
 #include "memory_monitor_sr.hpp"
 
+#ifndef HAVE_USR_INCLUDE_MALLOC_H
+#define HAVE_USR_INCLUDE_MALLOC_H
+#endif
+
+typedef struct mmon_metainfo
+{
+  uint64_t size;
+  int tag_id;
+  int magic_number;
+} MMON_METAINFO;
+
 namespace cubmem
 {
+  memory_monitor *mmon_Gl = nullptr;
+
   memory_monitor::memory_monitor (const char *server_name)
-    : m_server_name {server_name} {}
+    : m_server_name {server_name}
+  {
+    std::string magic_string ("MMON");
+    memcpy (&magic, magic_string.c_str (), sizeof (int));
+  }
 
-  static std::string memory_monitor::make_tag_name (const char *file, const int line)
+  size_t memory_monitor::get_alloc_size (char *ptr)
+  {
+    size_t ret;
+    size_t alloc_size = malloc_usable_size (ptr);
+    char *meta_ptr = ptr + alloc_size - MMON_ALLOC_META_SIZE;
+    MMON_METAINFO metainfo;
+
+    memcpy (&metainfo, meta_ptr, MMON_ALLOC_META_SIZE);
```

Comment:
I think we can remove the copy.
```suggestion
    const MMON_METAINFO * metainfo = (const MMON_METAINFO*) meta_ptr;
```

Reply:
Done.